contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c

```diff
@@ -374,6 +374,163 @@ tsql_check_param_readonly(const char *paramname, TypeName *typename, bool readon
+        {
+            // If not path is provided with use the standard path [$.columnName]
+            appendStringInfo(format_cols, "$.%s ", cd->elemName);
+        }
+        else
+        {
+            appendStringInfo(format_cols, "%s ", cd->elemPath);
+        }
+
+        // character types need to have the typmod appended to them
+        if((pg_strcasecmp(TypeNameToString(cd->elemType), "nvarchar") == 0 ||
+            (pg_strcasecmp(TypeNameToString(cd->elemType), "varchar") == 0 ||
+            (pg_strcasecmp(TypeNameToString(cd->elemType), "nchar") == 0 ||
+            pg_strcasecmp(TypeNameToString(cd->elemType), "char") == 0))))
+        {
+            int newTypMod = getNVarcharTypMod(cd->elemType);
+            char buffer [sizeof(int)*8+1];
+            sprintf(buffer, "%d", newTypMod);
+            appendStringInfo(format_cols, "%s(%s)", TypeNameToString(cd->elemType), buffer);
+        }
+        else
+        {
+            appendStringInfoString(format_cols, TypeNameToString(cd->elemType));
+        }
+
+        if((cd->asJson && pg_strcasecmp(TypeNameToString(cd->elemType), "nvarchar") == 0) && initialTmod == TSQLMaxTypmod)
+        {
+            appendStringInfoString(format_cols, " AS JSON");
+        }
+        else if(cd->asJson)
+        {
+            // AS JSON can only be used with nvarchar(max)
+            ereport(ERROR, (errcode(ERRCODE_SYNTAX_ERROR),
+                                errmsg("AS JSON in WITH clause can only be specified for column of type nvarchar(max)")));
+        }
+
+        jsonWithParams = lappend(jsonWithParams, makeStringConstCast(format_cols->data, -1, SystemTypeName("text")));
+        rf->coldeflist = lappend(rf->coldeflist, n);
+    }
+
+    fc = makeFuncCall(TsqlSystemFuncName("openjson_with"), jsonWithParams, COERCE_EXPLICIT_CALL, -1);
+    rf->functions = list_make1(list_make2(fc, NULL));
+    rf->alias = alias;
+    return (Node*) rf;
+}
+
+/*
+* Create a column definition node for the given column name and type
+* If the column type is a character type, we need to change the underlying typmod
+*/
+Node *
+createOpenJsonWithColDef(char* elemName, TypeName* elemType)
+{
+    ColumnDef *n = makeNode(ColumnDef);
+    n->colname = elemName;
+    if((pg_strcasecmp(TypeNameToString(elemType), "nvarchar") == 0 ||
+        (pg_strcasecmp(TypeNameToString(elemType), "varchar") == 0 ||
+        (pg_strcasecmp(TypeNameToString(elemType), "nchar") == 0 ||
+        pg_strcasecmp(TypeNameToString(elemType), "char") == 0))))
+    {
+        n->typeName = setNVarcharTypMod(elemType);
+    }
+    else
+    {
+        n->typeName = elemType;
+    }
+    n->inhcount = 0;
+    n->is_local = true;
+    n->is_not_null = false;
+    n->is_from_type = false;
+    n->storage = 0;
+    n->raw_default = NULL;
+    n->cooked_default = NULL;
+    n->collOid = InvalidOid;
+    n->constraints = NIL;
+    n->location = -1;
+    return (Node*) n;
+}
+
+TypeName *
+setNVarcharTypMod(TypeName *t)
```

Comment:
Given that this function is only called by openjson feature and used for all character types, a better naming is needed, e.g., `setCharTypmodForOpenjson`.